program/src/processor/set_authority.rs

```diff
@@ -13,9 +19,85 @@ use crate::instruction::{
 ///   1. `[s]` config_authority
 ///   2. `[]` new_authority
 pub fn process_set_authority(
-    _program_id: &Pubkey,
-    _ctx: Context<SetAuthorityAccounts>,
-    _authority_type: AuthorityType,
+    program_id: &Pubkey,
+    ctx: Context<SetAuthorityAccounts>,
+    authority_type: AuthorityType,
 ) -> ProgramResult {
+    // Accounts validation.
+
+    // 1. authority
+    // - must be a signer
+    // - must match the authority in the account (checked in the match statement below)
+
+    require!(
+        ctx.accounts.authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "authority"
+    );
+
+    // 2. account
+    // - owner must the stake program
+    // - must be initialized
+    // - must have an authority set
+
+    require!(
+        ctx.accounts.account.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "account"
+    );
+
+    let data = &mut ctx.accounts.account.try_borrow_mut_data()?;
+
+    match authority_type {
+        AuthorityType::Config | AuthorityType::Slash => {
+            let config = bytemuck::try_from_bytes_mut::<Config>(data)
+                .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+            require!(
+                config.is_initialized(),
+                ProgramError::UninitializedAccount,
+                "config"
+            );
+
+            // Asserts that the authority is set - only updates the authority jf there is
+            // one set and it matches the signing authority.
+            //
+            // TODO: do we fail if the authority is not set?
+
+            let current_authority = config.authority.into();
+
+            if let Some(current_authority) = current_authority {
+                require!(
+                    *ctx.accounts.authority.key == current_authority,
+                    StakeError::InvalidAuthority,
+                    "authority (config)"
+                );
+
+                match authority_type {
+                    AuthorityType::Config => {
+                        config.authority = OptionalNonZeroPubkey(*ctx.accounts.new_authority.key)
+                    }
+                    AuthorityType::Slash => {
+                        config.slash_authority =
+                            OptionalNonZeroPubkey(*ctx.accounts.new_authority.key)
+                    }
+                    _ => (), /* unreachable */
+                }
```

Comment:
What do you think about combining this and the encompassing match blocks? Can you make this a helper or closure that gets called into?

```rust
let config = bytemuck::try_from_bytes_mut::<Config>(data)
    .map_err(|_error| ProgramError::InvalidAccountData)?;

require!(
    config.is_initialized(),
    ProgramError::UninitializedAccount,
    "config"
);
```

This way, we can match on the `AuthorityType` once, and do away with the unreachable bit.

Reply:
You even roll a unified helper along these lines, up to you!

```rust
fn unpack_initialized_mut::<'a, T: Pod>(data: &'a mut [u8], name: &str) -> &'a mut T {
    let state = bytemuck::try_from_bytes_mut::<T>(data)
        .map_err(|_error| ProgramError::InvalidAccountData)?;

    require!(
        state.is_initialized(),
        ProgramError::UninitializedAccount,
        name
    );

    state
}
```

Reply:
I think we need a trait to do a function, we could use a macro instead.

Reply:
> I think we need a trait to do a function, we could use a macro instead.

`<'a, T: Pod>`

Reply:
If you think it's worth it, `solana-program` has such a trait:

https://github.com/anza-xyz/agave/blob/0ff5a4ba92dc68b446113a93547cd4affc7b0602/sdk/program/src/program_pack.rs#L5-L9

Reply:
Ooh nice, the macro is pretty nifty! Wfm.

Reply:
Good to know about the `IsInitialized` trait. I considered adding a similar trait, but then `Config` has `is_uninitialized` and it looked odd to not have it as part of the trait. But at the same time, `Stake` does not have the `is_uninitialized`. I then settle for the macro. ☺️